### ISSUE TITLE
[Feature] allow defer of sources

### ISSUE THREAD

**Issue by FDG4663** [dbt-labs/dbt-core#10912]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

In our organization we have a base layer dbt project (that creates the data warehouse) and then many reporting projects (that pull the the data warehouse and create models used in reporting). Since they are connected through sources and defer does not work with sources, we always have to hardcode our testing (deferred) data warehouse database into the dbt script wherever a source is used (deferral works perfectly for refs).

TL;DR:
Make `{{ source() }}` defer compatible (right now it changes just some parts of the database). Maybe even just by adding a fallback config option to switch to.

### Who will this benefit?

Everyone who uses defer in multi dbt project builds.

### Are you interested in contributing this feature?

I don't know the dbt code, but if I can help, let me know

**Comment by graciegoheen** [comment]
> Since they are connected through sources

Could you instead connect these projects as packages? Essentially install the base layer dbt project as a package in your downstream projects? 

Then, these would be connected by cross-project references: `{{ ref('project_or_package', 'model_name') }}`
